### ISSUE TITLE
Node.JS console app arguments

### ISSUE THREAD

**Issue by jgwinner** [MicrosoftDocs/visualstudio-docs#4584]
Hi:

I'm having a problem finding ANY documentation on how to configure debug arguments for node.js console apps.

I can successfully run the node.js console app by right-clicking and saying "set as startup item" and configuring a launch.vs.json file. For example, here's two different configurations, one a typical Node.JS server instance, the other cloned from Mean-Expert's Loopback generator command line project:

```
{
  "version": "0.2.1",
  "defaults": {},
  "configurations": [
    {
      "type": "nodeDbg",
      "name": "server.js",
      "project": "server\\server.js"
    },
    {
      "type": "nodeDbg",
      "name": "lb-cl-sdk.js",
      "project": "bin\\MyModelGen\\bin\\lb-cl-sdk.js"
    }
  ]
}
```

Yet, when I run the app, the console window flashes up so quickly as to be useless, and I need to pass command arguments to the runtime process.

If I try adding:
`
      "nodeArgs": "${workspaceFolder}..\\..\\..\\server\\server.js ${workspaceFolder}..\\..\\..\\client\\src\\app\\shared\\sdk\\ -w 'enabled'"`

I then get an error: 

One or more errors occurred.

Failed to launch debug adapter. Additional information may be available in the output window.

Cannot connect to runtime process, timeout after 10000 ms - (reason: Cannot connect to the target: connect ECONNREFUSED 126.0/0/1:5858).

I believe this is because adding the NodeArgs means that the typical --inspect-brk=5858 argument doesn't get added? I'd add myself, but don't know how to configure VS2019 to use a specfiic debug port.

I miss the C++ project parameters window.

Oddly, I have the same problem with Visual Studio Code, and it DOES add the argument.

I've searched for this information on the web and haven't gotten many hits, oddly (other than VS Code hits, where oddly I get the same problem).

        == John ==

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: dbb9adca-2c5c-edff-2c4e-4645579fb907
* Version Independent ID: 85cade65-f1b0-61d4-c69e-fb8ca94fefed
* Content: [Debug a JavaScript or TypeScript app - Visual Studio](https://docs.microsoft.com/en-us/visualstudio/javascript/debug-nodejs?view=vs-2019#feedback)
* Content Source: [docs/javascript/debug-nodejs.md](https://github.com/MicrosoftDocs/visualstudio-docs/blob/master/docs/javascript/debug-nodejs.md)
* Product: **visual-studio-windows**
* Technology: **vs-javascript**
* GitHub Login: @Mikejo5000
* Microsoft Alias: **mikejo**

**Comment by Mikejo5000** [comment]
@jgwinner This topic describes an "args" array for passing arguments to the debugger, though it's not clear if this will work for Node.js. @armanio123 for comment on what might or might not work here.

**Comment by jgwinner** [comment]
@Mikejo5000 Thanks Mike - basically, I need to pass arguments to the Node.JS program. It's a CLI app that 'does stuff' with Loopback. Without arguments, the program just prints the help screen and exits - not too useful.

Any help you can provide would be great - and update the docs. In VSXXXX it was pretty easy. I see Visual Studio is going the route of Visual Studio Code ... everything is hidden in JSON files instead of dialog boxes, and you've got to try to intellisense your way into meaningful JSON. The problem with that as a UI strategy is that it's frequently difficult to get any feedback on what an item is for, just it's name.

**Comment by Mikejo5000** [comment]
Sorry, forgot the link:
https://docs.microsoft.com/en-us/visualstudio/ide/customize-build-and-debug-tasks-in-visual-studio?view=vs-2019#configure-debugging-with-launchvsjson

**Comment by jgwinner** [comment]
Hmm ... I'm still having problems getting that to work. (which implies service request, but I think how to do this common operation should be documented).

I used 'args' but nothing gets passed through. The app reports no args, see screenshot. 

If I use 'NodeArgs' the args DO get passed through, but before the 'options' statement that has the debugger port, which has to be passed to Node (I wouldn't know that ahead of time, unless I can force a port ...hmm).

It shouldn't be this hard!

Not a 'hello world' example, but you can see 'args' is filled in, yet the app thinks it has 0 args

![image](https://user-images.githubusercontent.com/10425170/71933972-3be57d00-3158-11ea-8dc5-2eee818d2f97.png)

If I use 'NodeArgs' I get a connection refused, because the debugger port is passed in after the arguments, so node ignores the options.

![image](https://user-images.githubusercontent.com/10425170/71934534-87e4f180-3159-11ea-84f6-f3b85722a52e.png)

From the node debugger, you can see the out of place options:

[22:19:41.558 UTC] To client: "runInTerminal"({"kind":"external","title":"Node Debug Console","cwd":"C:\\c\\sanitized","args":["C:\\Program Files\\nodejs\\node.exe","bin\\MyModelGen\\bin\\lb-cl-sdk.js",".\\server\\server.js","client\\src\\app\\shared\\sdk\\","-w","\"enabled\"","--inspect-brk=5858","bin\\MyModelGen\\bin\\lb-cl-sdk.js"]}), timeout: 5000

**Comment by jgwinner** [comment]
Ah! That worked. My fix was:

      "debuggerPort": "1234",
      "nodeArgs": "--inspect-brk=1234 bin\\MyModelGen\\bin\\lb-cl-sdk.js .\\server\\server.js client\\src\\app\\shared\\sdk\\ -w \"enabled\""

The framework might (will) pass on a second --inspect-brk but it would be ignored, so the one we 'hard coded' will be the one that worked. Doing this I was able to hit "F10" and step into my code with VS2019! So long VS Code!

        == John ==

**Comment by Mikejo5000** [comment]
Thanks for the follow up! You are using an "Open Folder" solution, it looks like? I think that is required for launch.vs.jon.

**Comment by jgwinner** [comment]
I had originally, but there is a .sln from way back. Looks like it's folder open, I tried the .SLN and it didn't work.

With the above, rather frustratingly, it also doesn't leave the console window open. I checked both the debugger options and the Node.JS options; both are set to leave the windows open.

I'm able to get it to work more or less. VS2019 locks up occasionally when debugging, and the PC has to be hard reset (restart never works), but it might be Outlook related, not my above fix. (it's a dual CPU system ... which I think isn't designed to use this software).
